### PR TITLE
[NeMo-UX] Support `load_strictness`

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: llm

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by mikolajblaz — APPROVED** [review]
Thanks!

**Review by marcromeyn — CHANGES_REQUESTED** [review]
PL does this like [this](https://github.com/Lightning-AI/pytorch-lightning/pull/19404/files#diff-72f5c7707bc8f4033f0e5f644a8b2eda5d333084e91dfeaaa1368dcbdf24df02). So I feel like we should adopt the same design and not add a custom arg to our strategy.

**Comment by mikolajblaz** [comment]
> PL does this like [this](https://github.com/Lightning-AI/pytorch-lightning/pull/19404/files#diff-72f5c7707bc8f4033f0e5f644a8b2eda5d333084e91dfeaaa1368dcbdf24df02). So I feel like we should adopt the same design and not add a custom arg to our strategy.

If only we can propagate this flag to dist_checkpointing.load then using this flag would be ideal

**Comment by marcromeyn** [comment]
> > PL does this like [this](https://github.com/Lightning-AI/pytorch-lightning/pull/19404/files#diff-72f5c7707bc8f4033f0e5f644a8b2eda5d333084e91dfeaaa1368dcbdf24df02). So I feel like we should adopt the same design and not add a custom arg to our strategy.
> 
> If only we can propagate this flag to dist_checkpointing.load then using this flag would be ideal

Yeah, an example of how PL does that can be found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/b35f303561e0de06d559c9ec5322ffaa1a9017c0/src/lightning/pytorch/strategies/fsdp.py#L609).

**Comment by ashors1** [comment]
> > If only we can propagate this flag to dist_checkpointing.load then using this flag would be ideal
> 
> Yeah, an example of how PL does that can be found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/b35f303561e0de06d559c9ec5322ffaa1a9017c0/src/lightning/pytorch/strategies/fsdp.py#L609).

Revisiting this PR. I don't think this is quite what we want. It looks like PTL's `strict` is just passed to `load_model_state_dict`, which is called after the checkpoint is loaded using `dist_checkpointing.load`, so we're not able to propagate the value of `strict` to `dist_checkpointing.load`

File: nemo/lightning/_strategy_lib.py [hunk 1]
```diff
@@ -516,6 +516,19 @@ def load_model_state_dict(megatron_parallel, checkpoint: Mapping[str, Any], stri
     from megatron.core import parallel_state
     from megatron.core.dist_checkpointing.validation import StrictHandling, parse_strict_flag
 
+    ## convert from StrictHandling to bool for PTL
+    if os.environ.get("MCORE_STRICT_HANDLING") is not None:
```
Comment (mikolajblaz) [inline nemo/lightning/_strategy_lib.py:520]: Let's avoid such logic, it will be terrible to debug later on.

What's the reason this can't be passed by argument?

Comment (marcromeyn) [reply nemo/lightning/_strategy_lib.py:520]: I would prefer if it's set in a global var as opposed to a environment variable.

**Review by marcromeyn — APPROVED** [review]

File: nemo/lightning/_strategy_lib.py [hunk 2]
```diff
@@ -514,6 +514,19 @@ def get_safe(param_id):
...
 def load_model_state_dict(megatron_parallel, checkpoint: Mapping[str, Any], strict: bool = True) -> None:
     from megatron.core import parallel_state
+    from megatron.core.dist_checkpointing.validation import StrictHandling, parse_strict_flag
+
+    strict = parse_strict_flag(strict)
+    ## convert from StrictHandling to bool for PTL
+    if strict is not None and not isinstance(strict, bool):
+        strict_options = [
+            StrictHandling.RAISE_UNEXPECTED,
+            StrictHandling.RAISE_ALL,
```
Comment (mikolajblaz) [inline nemo/lightning/_strategy_lib.py:524]: I think we should also add `StrictHandling.ASSUME_OK_UNEXPECTED`?
Since this is the default one and we don't expect mismatches

File: nemo/lightning/_strategy_lib.py [hunk 3]
```diff
@@ -514,6 +514,19 @@ def get_safe(param_id):
...
+        ]
+        if strict in strict_options:
+            strict = True
+        else:
+            strict = False
```
Comment (mikolajblaz) [inline nemo/lightning/_strategy_lib.py:529]:
```suggestion
        strict = strict in strict_options
```

**Review by mikolajblaz — APPROVED** [review]

**Review by marcromeyn — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]
LGTM
